### ISSUE TITLE
 Re-frame side-causes (aka coeffect) as data?

### ISSUE THREAD

**Issue by myguidingstar** [roman01la/citrus#21]
Thanks for sharing this library. I don't see coeffect mentioned anywhere so I ask

**Comment by myguidingstar** [comment]
I've dug into source code, turns out the coeffect thing hasn't been implemented so I've done it myself.
The problem is this: if you want to use coeffect inside a controller's body, the reconciler must be passed implicitly as an argument to the controller. So, instead of:
```cljs
(defmethod some-controller :event-name
  [_event-name _params]
  ;; impure body
  {:state (calculate-stuff (rand-int 42))})
```
controller will become
```cljs
(defmethod some-controller :event-name
  [reconciler _event-name _params]
  ;; it's pure and can be mocked when testing
  (let [my-number (citrus/get-cause reconciler :my-random-number)]
    {:state (calculate-stuff my-number)}))

(defonce reconciler
  (citrus/reconciler
    {:state           (atom)
     :controllers     {:foo some-controller}
     ;; new stuff :D
     :cause-handlers  {:my-random-number #(rand-int 42)}}))
```

**Comment by myguidingstar** [comment]
sure this will break existing code. If you think it's worth it, I'll make an PR

**Comment by roman01la** [comment]
@myguidingstar I'm not much into re-frame, could you please tell more about co-effects? In my understanding it is a mechanism to make event handlers pure with respect to its dependencies, am I right? What problems this is solving? I can immediately think about testing only.

**Comment by roman01la** [comment]
it could be something like this
```clojure
(defmethod control :fetch-profiles [_ [account-id] state coeffects]
  {:state (assoc state :loading? true)
   :http  {:method  :get
           :url     "/api/profiles"
           :params  {:account-id account-id
                     :token      (:auth-token coeffects)}
           :on-done [:profile :fetch-profiles-done]}})
```

But then we need a way to register specific co-effect handlers to those controllers. I believe it could be done with a macro that returns `defmethod` and does a little bit of magic to register co-effects

```clojure
(defn local-store [key]
  (js/localStorage.getItem key))

(citrus/defhandler control :fetch-profiles {:cofx {:local-store #(cofx/local-store "auth-token")}}
  [_ [account-id] state coeffects]
  {:state (assoc state :loading? true)
   :http  {:method  :get
           :url     "/api/profiles"
           :params  {:account-id account-id
                     :token      (:auth-token coeffects)}
           :on-done [:profile :fetch-profiles-done]}})
```

**Comment by roman01la** [comment]
I've pushed initial implementation here https://github.com/roman01la/citrus/tree/coeffects

**Comment by myguidingstar** [comment]
the PR looks neat. However I don't quite get the reason for using metadata here. And how will you mock co-effect handlers when testing?

**Comment by roman01la** [comment]
yeah, I've changed that already. for now it looks like this

```clojure
(cofx/defhandler control :load
  {:cofx [[:local-storage :count]]} ;; <- a vector of co-effects [cofx-d & args]
  [_ [key] _ coeffects]
  {:state (-> coeffects :local-storage int)})

(defonce reconciler
         (citrus/reconciler
           {:state           (atom {})
            :controllers     {:counter control}
            :effect-handlers {:local-storage local-storage}
            :co-effects      {:local-storage #(js/localStorage.getItem (name %))}})) ;; <- co-effects handlers
```

Co-effects are stored as meta data on a defmulti so it can be retrieved later, in reconciler, when executing events. It's kind of a local registry for co-effects, instead of global one as in re-frame.

**Comment by myguidingstar** [comment]
Yeah I know it must be stored somewhere to be retrieved later. Sorry, I should have made better question. I was asking why metadata instead of global registry, but It makes sense now.
I think you can merge it now.

**Comment by myguidingstar** [comment]
Btw, in the example, multiple unnamed symbols `_` next to each other look scary to new api consumers like me. Please keep the var name semantic. Instead of:
```clj
(defmethod control :dec [_ _ counter]
 ...)
```
you can:
```
(defmethod control :dec [_event _args counter]
  ...)
```
but you've done enough for this PR. The thing mentioned above should be in another PR

**Comment by roman01la** [comment]
published experimental support  in `3.1.0-SNAPSHOT`
